doc/examples/filters/plot_j_invariant.py

```diff
@@ -0,0 +1,294 @@
+========================================
+
+In this example, we show how to find an optimally calibrated
+version of any denoising algorithm.
+
+"""
+
+#####################################################################
+# Calibrating a wavelet denoiser
+
+import numpy as np
+from matplotlib import pyplot as plt
+from matplotlib import gridspec
+
+from skimage.data import chelsea, hubble_deep_field
+from skimage.measure import compare_mse as mse
+from skimage.measure import compare_psnr as psnr
+from skimage.restoration import (calibrate_denoiser, calibrate_denoiser_search,
+                                 invariant_denoise, denoise_wavelet,
+                                 denoise_tv_chambolle, denoise_nl_means,
+                                 estimate_sigma)
+from skimage.util import img_as_float, random_noise
+from skimage.color import rgb2gray
+
+
+image = img_as_float(chelsea())
+sigma = 0.3
+noisy = random_noise(image, var=sigma ** 2)
+
+sigma_range = np.arange(0.05, 0.5, 0.05)
+
+parameter_ranges = {'sigma': np.arange(0.1, 0.4, 0.02),
+                    'wavelet': ['db1', 'db2', 'sym9'],
+                    'convert2ycbcr': [True, False],
+                    'multichannel': [True]}
+
+default_output = denoise_wavelet(noisy, multichannel=True)
+
+calibrated_denoiser = calibrate_denoiser(noisy,
+                                         denoise_wavelet,
+                                         parameter_ranges,
+                                         multichannel=True)
+
+calibrated_output = calibrated_denoiser(noisy)
+
+fig, axes = plt.subplots(1, 3, sharey=True, figsize=(15, 5))
+
+for ax, img, title in zip(axes,
+                          [noisy, default_output, calibrated_output],
+                          ["Noisy Image", "Denoised (Default)",
+                           "Denoised (Calibrated)"]):
+    ax.imshow(img)
+    ax.set_title(title)
+    ax.set_yticks([])
+    ax.set_xticks([])
+
+#####################################################################
+# The Self-Supervised Loss and J-Invariance
+# =========================================
+# The key to this calibration method is the notion of J-invariance. A denoising
+# function is J-invariant if the prediction it makes for each pixel does
+# not depend on the value of that pixel in the original image. The prediction
+# for each pixel may instead use all the relevant information contained in the
+# rest of the image, which is typically quite significant. Any function
+# can be converted into a J-invariant one using a simple masking procedure,
+# as described in [1].
+#
+# The pixel-wise error of a J-invariant denoiser is uncorrelated
+# to the noise, so long as the noise in each pixel is independent.
+# Consequently, the average difference between the denoised image and the
+# noisy image, the _self-supervised loss_, is the same as the
+# difference between the denoised image and the original clean image, the
+# _ground-truth loss_ (up to a constant).
```

Comment:
same here

Reply:
Thank you for the thoughtful comments, @emmanuelle.

> * at the moment when a set of parameters is provided, the optimum is computed by grid search / brute force. Would it be interesting to use a "real" optimization algorithm such as `scipy.optimize.minimize` (with a fairly lenient tolerance)? On a related note, do we expect the loss to be a convex function of hyperparameters or not? If it's a very complicated function, the brute force might be better...

I like this idea. The runtime will just be proportional to the number of parameter values tested, but can also be brought down by working on a representative subset of the image.

For every search I have done, the function has been convex in the ordinal hyperparameters (like`sigma` for wavelets, `h` and `patch_size` and `patch_distance` for NL means), at least in any neighborhood of the optimum. (I'm not sure what happens if you start off wrong by an order of magnitude.) No guarantees about the categorical parameters (which family of wavelets or which thresholding, for example).

> * A lot of the for loops are embarrassingly parallel and could be accelerated with dask, see for example https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_cycle_spin.py#L166

That would be cool! I haven't used dask before. To speed up the for loops in `invariant_denoise` and `calibrate_denoiser_search`, would I need to make them into list comprehensions as in the example?